feat: add global specification watcher

imabp opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

- Adds Global Watcher for Specification

![image](https://user-images.githubusercontent.com/53480076/153071983-a768fb0d-b469-42f3-ba0f-57a5e8eb4dd9.png)


**Related issue(s)**
#73 
<!-- If you refer to a particular issue, provide its number, othewise, remove this section.
For example, `Resolves #123`, `Fixes #43`, or `See also #33`. The `See also #33` option will not automatically close the issue after the PR merge. -->

imabp commented:
I guess there is an error with, test case for watch mode. 
The fs stream is not getting closed. I will try to create a mock stream and then close it from tests
Apologies, the test ran for 4 hours.

derberg commented on src/commands/validate.ts:
```diff
@@ -4,21 +4,29 @@ import Command from '../base';
   static flags = {
-    help: flags.help({ char: 'h' })
+    help: flags.help({ char: 'h' }),
+    watch: flags.boolean({
```
maybe we could do it more generic, so we do not have to repeat the description of the flag everywhere. So it is basically not `flags.boolean` but `flags.watch`? @Souvikns @magicmatatjahu @boyney123 thoughts?

imabp replied:
@derberg  Do you mean, building a wrapper over `flags.boolean`

magicmatatjahu replied:
@imabp Yep, but I think that Łukasz thinks about creating reusable flag and then reuse it in other commands. You can check documentation how do that in the ocliff https://oclif.io/docs/flags#custom-flags

imabp replied:
Sure, thank you @magicmatatjahu 
I was looking at that only.

Just figuring out, how do I pass on the `current command`,`commandName` and `filepath` to the `specWatcher.`

imabp replied:
Thanks @magicmatatjahu for the resource. I have refactored the watcher.  Please review @derberg @magicmatatjahu

imabp replied:
I am also figuring it out, how to make it more better, just like help command, need some time to explore those parts.

derberg reviewed (changes requested):
Please also enable watcher for diff

@Souvikns @boyney123 @magicmatatjahu wanna have a look?

imabp commented:
Hey @derberg `watch` flag is added to diff :)
Kindly review :)

magicmatatjahu commented on src/globals.ts:
```diff
@@ -0,0 +1,27 @@
+const GreenLog = chalk.hex('#00FF00');
+const OrangeLog = chalk.hex('#FFA500');
+
+export const specWatcher = (filePath: string, handler: any, handlerName: string) => {
```
```suggestion
export const specWatcher = (filePath: string, handler: Command, handlerName: string) => {
```

Please import `Command` class type :)

imabp replied:
Oops @magicmatatjahu my recent commit on local machine, already fixes this. 😅
Thanks for going through that again :)

magicmatatjahu commented on src/globals.ts:
```diff
@@ -0,0 +1,27 @@
+const OrangeLog = chalk.hex('#FFA500');
+
+export const specWatcher = (filePath: string, handler: any, handlerName: string) => {
+  const CHOKIDAR_CONFIG = {
+    awaitWriteFinish: true
+  };
+  const WATCH_MESSAGES = {
+    logOnStart: (filePath: string) => console.log(GreenLog(`Watching AsyncAPI file at ${filePath}`)),
+    logOnChange: (handlerName: string) => console.log(OrangeLog(`Change detected, running ${handlerName}`)),
+  };
```
Please move that arrays outside `specWatcher` function because you don't need to recreate them on each evaluation of `specWatcher` function :)

magicmatatjahu reviewed (changes requested):
Awesome! I leave some comments, please refer to them.

magicmatatjahu reviewed (changes requested):
@imabp Good, but you have to handle that one code smell :) https://sonarcloud.io/project/issues?id=asyncapi_cli&open=AX7bDmqtPXvNfWF5cO0t&pullRequest=220&resolved=false&types=CODE_SMELL

Also please refer to my and Łukasz suggestion: https://github.com/asyncapi/cli/pull/220/files#r802512653

imabp commented:
Hey @magicmatatjahu the code smell is fixed :)

derberg commented on src/commands/diff.ts:
```diff
@@ -125,6 +128,10 @@ export default class Diff extends Command {
         err: error,
       });
     }
+    if (watchMode) {
```
I think we also need conditions for `firstDocument.getFilePath()` and `secondDocument.getFilePath()` right? because if they are passed as url, then watched doesn't make much sense.

This also makes me think that we most probably should probably log a message to the use that in case AsyncAPI file path is passed as URL than watching doesn't start and we say `sorry boss, not this time` 😄

magicmatatjahu replied:
Last thing, rest is good :)

We should move that `if` at the begging of logic like in `validation` command, after 67 line :)

derberg commented on src/flags.ts:
```diff
@@ -0,0 +1,6 @@
+import { flags } from '@oclif/command';
+
+export const watchFlag = flags.boolean({
```
it works fine without `Flags.build`? https://oclif.io/docs/flags#custom-flags 🤔

imabp replied:
@magicmatatjahu , `flags.build`, expects a string value as input to `--watch`, but `--watch` is a boolean.
So I had to go with this

magicmatatjahu replied:
Yeah, you're right... They force the string type... 🤦🏼  https://github.com/oclif/core/blob/main/src/parser/flags.ts#L21 Ok, leave it as it is.

Please propagate suggestion from Łukasz (about chokadir and fix text in logs) and ping me for next review :)

derberg reviewed (changes requested):
OH man, what an improvement, it looks much cleaner now. I left only 2 comments

imabp commented:
Hey @derberg @magicmatatjahu 
Please check the latest changes, of auto-disabling watch mode. Here is a screenshot

![image](https://user-images.githubusercontent.com/53480076/153386824-28ebd93c-f9da-4855-ae88-d80433d60141.png)

derberg commented on src/globals.ts:
```diff
@@ -0,0 +1,29 @@
+export const WATCH_MESSAGES = {
+  logOnStart: (filePath: string) => console.log(GreenLog(`Watching AsyncAPI file at ${filePath}\n`)),
+  logOnChange: (handlerName: string) => console.log(OrangeLog(`Change detected, running ${handlerName}\n`)),
+  logOnAutoDisable: (docVersion: 'old' | 'new') => console.log(RedLog(`Watch mode for ${docVersion} file was not enabled..\n`), OrangeLog('INFO: Watch works only with files from local file system\n'))
```
```suggestion
  logOnAutoDisable: (docVersion: 'old' | 'new') => console.log(OrangeLog(`Watch mode for ${docVersion} file was not enabled..\n`), OrangeLog('INFO: Watch works only with files from local file system\n'))
```

I think we should use red only for errors/failures, and here it is just an info really

derberg commented on src/globals.ts:
```diff
@@ -0,0 +1,29 @@
+export const OrangeLog = chalk.hex('#FFA500');
+export const RedLog = chalk.hex('#FF0000');
+const CHOKIDAR_CONFIG = {
+  // awaitWriteFinish: true
```
what about this one?

imabp replied:
I was bit confused whether to enable this or not. 
Usually specification files, having relatively larger size than our sample ones, may take some time to save.

When `awaitWriteFinish` is `true`
So the watcher triggers callback after 2 seconds, till the changes get saved. 
Also, that means, we dont get the log message instantaneously, wait for 2 seconds, after the change has been made.

imabp replied:
I have appended the explanation to the comment in the latest changes :)

derberg commented on src/globals.ts:
```diff
@@ -0,0 +1,29 @@
+  logOnAutoDisable: (docVersion: 'old' | 'new') => console.log(RedLog(`Watch mode for ${docVersion} file was not enabled..\n`), OrangeLog('INFO: Watch works only with files from local file system\n'))
+};
+
+export const specWatcher = (filePath: string, handler: Command, handlerName: string) => {
```
should not logging of `logOnAutoDisable` actually be handled here?

imabp replied:
Yes @derberg  thanks for that refactor idea
I will try to refactor the above logic into specWatcher only.

derberg reviewed (changes requested):
looking good, except for one followup comments I guess we are still left with an idea on testing watcher?

I'm also wondering if we do not need unit tests for commands where watcher is used. Like what happens for example when user passes `--watch` and both AsyncAPI references are URLs. Do we perform validation/diff> IMHO yes, but then we need to have it tested imho as well

imabp commented on src/globals.ts:
```diff
@@ -0,0 +1,29 @@
+
+export const GreenLog = chalk.hex('#00FF00');
+export const OrangeLog = chalk.hex('#FFA500');
+export const RedLog = chalk.hex('#FF0000');
```
Yes, its private now :)

imabp commented:
@derberg I guess, we are handling these edge cases really well in diff and also in validate(now)
1. Both are URLs (in diff) 
    - Result: Do not call the watcher
2. One is URL and another is Local File ( in diff )
    - Result: Call the watcher for the local file only.
3. The file is URL in validate 
     - Result: Do not  call the watcher.
     

**The problem with test case**     

I think, if we are writing test case for watcher, we need to have a different approach or test case, just for watcher, because it has a requirement to create a mock file stream and then when the tests complete, we need to kill that mock file stream.

Why?
As watcher opens a file stream and watches it. But Mocha doesn't close all the file streams, when tests are complete.
Will be glad, if @fmvilas wants to share his suggestion on this specific aspect of Mocha.

imabp commented:
@magicmatatjahu I tried building the --watch flag using `flags.build`, but the library itself  accepts a string as an input for `--watch` flag, but watch is boolean. That's why,  for boolean, I had to do a different workout

Attaching a screenshot of the same, if you see in the popup the type of flag `input` is string, but we need boolean, that's why.

Do let me know, if something I am missing out here. 

![buildOclif](https://user-images.githubusercontent.com/53480076/153443794-5b1640b4-d769-4f34-87e6-dd9145fd5d91.png)

derberg reviewed (dismissed):
code is clear.

I let @magicmatatjahu continue discussion and flag builder

there is just a problem with the current implementation that when change is detected, then handler is called again. All good but it basically creates another "instance" of chokidar watcher, and then, when another change is detected, then 2 instances create another 2 instances of chokidar, and then 4 create 4, and then you basically see a lot of the same log messages. 

So you just need to make sure that validate and diff call watcher only once. I hope I explained it well

@magicmatatjahu I approve from my side, because I'm out next 2 weeks and I do not want to block merge here. There are just few minor stuff from me left. For now I will marke it as /dnm

magicmatatjahu commented:
@imabp I past my comment from one thread because I see that github didn't show it at the bottom. Comment about `Flags.builder`.

Yeah, you're right... They force the string type... 🤦🏼 https://github.com/oclif/core/blob/main/src/parser/flags.ts#L21 Ok, leave it as it is.

Please propagate suggestion from Łukasz (about chokadir and fix text in logs) and ping me for next review :)

imabp commented on src/globals.ts:
```diff
@@ -0,0 +1,39 @@
+  spec:Specification,
+  handler:Command,
+  handlerName: string,
+  docVersion?: 'old' | 'new' | ''
```
Thank you, its fixed now 😊

imabp commented:
@magicmatatjahu @derberg 

1. FIXED: Chokidar initiating multiple instances is fixed now
2. FIXED: Make logs look consistent by adding conditional logs to watcher based on `docVersion`

imabp commented:
@magicmatatjahu  please review :)

magicmatatjahu commented on src/globals.ts:
```diff
@@ -0,0 +1,48 @@
+const WATCH_MESSAGES = {
+  logOnStart: (filePath: string) => console.log(GreenLog(`Watching AsyncAPI file at ${filePath}\n`)),
+  logOnChange: (handlerName: string) => console.log(OrangeLog(`Change detected, running ${handlerName}\n`)),
+  logOnAutoDisable: (docVersion: 'old' | 'new' | '' = '') => console.log(OrangeLog(`Watch mode for${ConditionalNestedLogs.logOnAutoDisable(docVersion)}file was not enabled.`), OrangeLog('\nINFO: Watch works only with files from local file system\n'))
```
You can simplify it:

```suggestion
  logOnAutoDisable: (docVersion: 'old' | 'new' | '' = '') => console.log(OrangeLog(`Watch mode for ${docVersion || 'AsyncAPI'} file was not enabled.`), OrangeLog('\nINFO: Watch works only with files from local file system\n'))
```

magicmatatjahu commented on src/globals.ts:
```diff
@@ -0,0 +1,48 @@
+  // awaitWriteFinish: true // Used for large size specification files.
+
+};
+const ConditionalNestedLogs = {
+  logOnAutoDisable: (docVersion: 'old' | 'new' | '' = '') => docVersion === '' ? (' ' + 'AsyncAPI' + ' ') : (` ${docVersion} `)
+};
```
You don't need that function :) Check another comment

```suggestion
```

magicmatatjahu commented on src/globals.ts:
```diff
@@ -0,0 +1,48 @@
+  spec: Specification,
+  handler: Command,
+  handlerName: string,
+  docVersion?: 'old' | 'new' | ''
```
I think that signature for that field should be:

```suggestion
  docVersion?: 'old' | 'new';
```

because optional field say that it can be also `undefined`, empty string is unnecessary, because it can be treated as undefined.

magicmatatjahu commented on src/globals.ts:
```diff
@@ -0,0 +1,48 @@
+}
+
+export const specWatcher = (params: specWatcherParams) => {
+  if (!params.spec.getFilePath()) { return WATCH_MESSAGES.logOnAutoDisable(params.docVersion); }
+
+  const filePath = params.spec.getFilePath() as string;
+  try {
+    WATCH_MESSAGES.logOnStart(filePath);
+    if (!IS_CHOKIDAR_INSTANCE_RUNNING) {
```
I think that `if (!IS_CHOKIDAR_INSTANCE_RUNNING) {` condition should be checked at the begging of that function to avoid unnecessary operation between execution of the `run` function:

 ```js
  if (IS_CHOKIDAR_INSTANCE_RUNNING) return; 
  if (!params.spec.getFilePath()) { return WATCH_MESSAGES.logOnAutoDisable(params.docVersion); }
```

magicmatatjahu reviewed (commented):
Great! I have some suggestions :)

imabp commented:
Hello @magicmatatjahu, just to give it a consistent look, I have added the watch start message after a new change. 
Feel free to review :) Attaching a screenshot
![image](https://user-images.githubusercontent.com/53480076/153850963-ab49273e-2776-489c-845f-40c007107e88.png)

magicmatatjahu commented on src/commands/diff.ts:
```diff
@@ -125,8 +128,11 @@ export default class Diff extends Command {
       });
     }
+    if (watchMode) {
+      specWatcher({spec: firstDocument, handler: this, handlerName: 'diff', docVersion: 'old'});
+      specWatcher({spec: secondDocument, handler: this, handlerName: 'diff', docVersion: 'new'});
```
Here we have a serious error. As we create a watcher only once, a second watcher will not be created at all. We need to be able to create multiple watchers. We suggest to add a label to the watcher and then create/check if it exists (we can use map to store the watchers):

```ts
if (watcherMap.has({Label})) {
  return
}

...
watcherMap.set({Label}, {Watcher})
```

imabp replied:
Yeah, I noticed this. Just built a nice store for it :)

magicmatatjahu commented on src/globals.ts:
```diff
@@ -0,0 +1,63 @@
+  logOnAutoDisable: (docVersion: 'old' | 'new' | '' = '') => console.log(OrangeLog(`Watch mode for ${docVersion || 'AsyncAPI'} file was not enabled.`), OrangeLog('\nINFO: Watch works only with files from local file system\n'))
+};
+
+type CHOKIDAR_INSTANCE_STORE = Record<string, boolean>
+
+const IS_CHOKIDAR_INSTANCE_RUNNING: CHOKIDAR_INSTANCE_STORE = {
+  _default: false,
+};
+
+export type specWatcherParams = {
+  spec: Specification,
+  handler: Command,
+  handlerName: string,
+  label?: string,
+  docVersion?: 'old' | 'new';
+}
+
+const getLabelValue = (label: string): boolean => {
+  return IS_CHOKIDAR_INSTANCE_RUNNING[label as string];
+};
+const setLabelValue = (label: string, val: boolean) => {
+  IS_CHOKIDAR_INSTANCE_RUNNING[label as string] = val;
+};
+
+export const specWatcher = (params: specWatcherParams) => {
+  if (!params.spec.getFilePath()) { return WATCH_MESSAGES.logOnAutoDisable(params.docVersion); }
+  if (!params.label && getLabelValue('_default')) { return; }
+  if (params.label && getLabelValue(params.label as string)) { return; }
```
Why do you want to use normal object, not `Map`? I think that it's better here. Really, please consider to use `Map`, Then you won't need to write that `getLabelValue`, `setLabelValue` functions because you have `.set`, `.get` in the `Map` prototype.

Also, if you don't have `params.label` you don't need to write separate condition for `_default` label but you can also do:

```ts
const hasLabel = getLabelValue(params.label || '_default');
// and
setLabelValue(params.label || '_default', true);
```

imabp replied:
Sure please review the latest changes :)

magicmatatjahu reviewed (changes requested):
Sorry for that long review, but we need to improve some things :)

magicmatatjahu reviewed (changes requested):
I know that there're another problems from my side but it's important. I see two bugs:

- more important: when I validate some spec with watch flag and then I e.g. remove info.title field which is important I get error (which is good) but also watching closes. As a user, if I enable watch flag I should have possibility to write spec and if I write invalid spec I should continue my work. Check example

![image](https://user-images.githubusercontent.com/20404945/154483432-d01ea341-8f76-4853-91ef-258a5033fdba.png)

  We need to change logic for `validate` command and try to handle validation with watching in another way, that error shouldn't close chokadir.

- another problem: when I provide some changes in the watched file I always see `Watching AsyncAPI file at ./sample.yaml` on every change. We should remove `40` line from `specWatcher` function:

![image](https://user-images.githubusercontent.com/20404945/154483695-7eb9ee58-0c56-4721-ae4e-18c7f1352749.png)

magicmatatjahu commented:
@imabp I would like to inform you that this feature probably will be needed for these PRs:

- https://github.com/asyncapi/cli/pull/221
- https://github.com/asyncapi/cli/pull/188

I don't want to rush you, but please have it in mind. For `generate` command it will be needed that fix for problem described in the https://github.com/asyncapi/cli/pull/220#pullrequestreview-885904531 Of course I can help you if you don't think how to solve that problem :)

imabp commented:
@magicmatatjahu , I am unable to reproduce the error of auto-closing of watcher, as you mentioned. 
Can you please recheck it at your end. 

![image](https://user-images.githubusercontent.com/53480076/154682521-4a58c5d6-b49a-4d55-8df8-880d35ec6220.png)

magicmatatjahu commented:
@imabp Hmm, I can reproduce it from the latest commit:
- create `sample.yaml` file with content from https://raw.githubusercontent.com/asyncapi/spec/master/examples/streetlights-kafka.yml
- build CLI and run `./bin/run validate ./sample.yaml -w`
- remove `title` from `info` object 

and I see this:
![image](https://user-images.githubusercontent.com/20404945/154683244-54686442-73db-4562-9421-67a63653f92c.png)

I use the MacOS, but I don't think so that there are some problems from system perspective. Also I use `v16.13.2` od NodeJS.

magicmatatjahu commented:
@imabp Ok I see that from `v15` of NodeJS unhandled Promise exceptions are treated with strict flag, not as warn/optional flag so with v16 my process closes but you probably use <v15 so your process is still alive - check this https://developer.ibm.com/blogs/nodejs-15-release-blog/

imabp commented:
Umm. @magicmatatjahu, okay so how do we proceed forward on this?
I crosschecked my node version is 14.19.0 

Do we run the start command having non-strict method? or something else

imabp commented:
Shall we run the watcher again, in the `catch(err)` block

imabp commented:
I think we should go with the first approach 
>We should have that same behaviour in all NodeJS versions (supported versions of course) and I prefer to handle all unhandledRejection by this code:
>```ts
> process.on('unhandledRejection', (reason, promise) => {
>  // throw error and stop process
> });```

May be have it ready in the `src/globals.ts` itself? @magicmatatjahu

magicmatatjahu commented:
> I think we should go with the first approach

but handle `unhandledRejection` we should make in separate PR as I wrote :)

imabp commented:
Sure @magicmatatjahu here is the new issue, we can work on that. :) 
https://github.com/asyncapi/cli/issues/234

imabp commented:
Just a question, @magicmatatjahu then shall we proceed with this PR, and then address the issue of handling errors #234 ? 
Or
Work on #234 and let this PR be on hold. ?

magicmatatjahu commented:
@imabp First complete that PR and then start working on mentioned issue :) You can propagate changes from that comment and I will accept it. https://github.com/asyncapi/cli/pull/220#issuecomment-1044498676

imabp commented:
I have fixed the second error as you mentioned @magicmatatjahu and is now reflected in the latest changes
>another problem: when I provide some changes in the watched file I always see Watching AsyncAPI file at ./sample.yaml on every change. We should remove 40 line from specWatcher function:

![image](https://user-images.githubusercontent.com/53480076/154692413-8057b29e-58eb-43e8-85ba-fa17d590f22c.png)

Here are the checks
![image](https://user-images.githubusercontent.com/53480076/154692551-0e3fcaf2-2909-476d-9179-ece5a4364d96.png)

magicmatatjahu commented:
@imabp Probably we don't understand each other 😆  I had in mind this one:

![image](https://user-images.githubusercontent.com/20404945/154693314-ba0c49c4-3e0f-4238-bd3c-775f2d4926d2.png)

imabp commented:
Ohh I got it.. 😅 . .Its really a learning to converse on PRs 😄 specially on such highly-critical issues
Implementing inside the `catch(err)` block, sure I will go through that..

imabp commented:
Hey @magicmatatjahu 
Everything seems to work well now 😄
Checked on node LTS 
![image](https://user-images.githubusercontent.com/53480076/154698611-c0c6e9cc-263c-4ac5-bce9-bf3cac731215.png)

magicmatatjahu reviewed (commented):
Last thing :)

magicmatatjahu reviewed (approved):
LGTM! However I see little "bug" when I run `diff` command. When I have invalid spec I cannot recognize in which file I have errors (in new or old?):

![image](https://user-images.githubusercontent.com/20404945/154952695-50f29e04-2048-462a-b84b-358adef62930.png)

We should improve logs in the followup PR @imabp Could you handle that? We need that PR for another commands so I don't want to block development with rejection.

magicmatatjahu commented:
/rtm

imabp commented:
Yes sure!! We need a proper logging control center 😅 or an interface

I am creating a issue #236  known as https://github.com/asyncapi/cli/issues/236
We can work on this after fixing #234

imabp commented:
@all-contributors please add @imabp for code

imabp commented:
I gave it a try, and it worked out xD 😂😂
